src/main/java/ch/njol/skript/sections/EffSecFireworkLaunch.java

```diff
@@ -0,0 +1,166 @@
+
+		public FireworkLaunchEvent(Firework firework) {
+			this.firework = firework;
+		}
+
+		public Firework getFirework() {
+			return firework;
+		}
+
+		@Override
+		public HandlerList getHandlers() {
+			throw new IllegalStateException();
+		}
+	}
+
+	static {
+		Skript.registerSection(EffSecFireworkLaunch.class, "(launch|deploy) [[a] firework [with effect[s]]] %fireworkeffects% at %locations% [([with] (duration|power)|timed) %-number%]");
+		EventValues.registerEventValue(FireworkLaunchEvent.class, Firework.class, new Getter<Firework, FireworkLaunchEvent>() {
+			@Override
+			public Firework get(FireworkLaunchEvent fireworkLaunchEvent) {
+				return fireworkLaunchEvent.getFirework();
+			}
+		}, EventValues.TIME_NOW);
+	}
+
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private Expression<FireworkEffect> effects;
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private Expression<Location> locations;
+	@Nullable
+	private Expression<Number> lifetime;
+
+	@Nullable
+	private Trigger trigger;
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs,
+						int matchedPattern,
+						Kleenean isDelayed,
+						SkriptParser.ParseResult parseResult,
+						@Nullable SectionNode sectionNode,
+						@Nullable List<TriggerItem> triggerItems) {
+		effects = (Expression<FireworkEffect>) exprs[0];
+		locations = (Expression<Location>) exprs[1];
+		lifetime = (Expression<Number>) exprs[2];
+
+		if (sectionNode != null) {
+			trigger = loadCode(sectionNode, "fireworklaunch", FireworkLaunchEvent.class);
+		}
+
+		return true;
+	}
+
+	@Override
+	@Nullable
+	@SuppressWarnings({"unchecked", "rawtypes"})
+	protected TriggerItem walk(Event e) {
+		Object localVars = Variables.copyLocalVariables(e);
+
+		Consumer<? extends Firework> consumer;
+		if (trigger != null) {
+			consumer = o -> {
+				FireworkLaunchEvent fireworkLaunchEvent = new FireworkLaunchEvent(o);
+				// Copy the local variables from the calling code to this section
+				Variables.setLocalVariables(fireworkLaunchEvent, localVars);
+				trigger.execute(fireworkLaunchEvent);
+				Variables.setLocalVariables(e, localVars); // Carry over variable changes to the rest of the main trigger
+			};
+		} else {
+			consumer = null;
+		}
+
+		Number power = lifetime != null ? lifetime.getSingle(e) : 1;
+		if (power == null)
+			power = 1;
+		for (Location location : locations.getArray(e)) {
+			Firework firework = location.getWorld().spawn(location, Firework.class, (Consumer) consumer);
+			FireworkMeta meta = firework.getFireworkMeta();
+			meta.addEffects(effects.getArray(e));
+			meta.setPower(power.intValue());
+			firework.setFireworkMeta(meta);
+		}
+
+		return super.walk(e, false);
+	}
+
+	@Override
+	public String toString(@Nullable Event e, boolean debug) {
+		return "Launch firework(s) " + effects.toString(e, debug) +
+			" at location(s) " + locations.toString(e, debug) +
+			" timed " + lifetime != null ? lifetime.toString(e, debug) : "1";
```

Comment:
```suggestion
			" timed " + (lifetime != null ? lifetime.toString(e, debug) : "1");
```
I don't know that it necessarily matters, but parentheses here would be better